File: license-tracker-backend/src/main/java/com/prodapt/license_tracker_backend/service/AuditLogService.java
// src/main/java/com/prodapt/license_tracker_backend/service/AuditLogService.java
package com.prodapt.license_tracker_backend.service;

import com.prodapt.license_tracker_backend.dto.AuditLogResponse;
import com.prodapt.license_tracker_backend.dto.CreateAuditLogRequest;
import com.prodapt.license_tracker_backend.entities.enums.AuditAction;
import com.prodapt.license_tracker_backend.entities.enums.EntityType;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDateTime;

public interface AuditLogService {

    void log(CreateAuditLogRequest request);

    void log(Long userId, String username, EntityType entityType,
             String entityId, AuditAction action, String details);

    Page<AuditLogResponse> getAllAuditLogs(Pageable pageable);

    Page<AuditLogResponse> getAuditLogsByUser(Long userId, Pageable pageable);

    Page<AuditLogResponse> getAuditLogsByEntity(EntityType entityType, String entityId, Pageable pageable);

    Page<AuditLogResponse> getAuditLogsByFilters(
            EntityType entityType,
            AuditAction action,
            Long userId,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable);

    Page<AuditLogResponse> searchAuditLogs(String searchTerm, Pageable pageable);

    AuditLogResponse getAuditLogById(Long logId);

    /**
     * NEW: Search by license key
     */
    Page<AuditLogResponse> searchByLicenseKey(String licenseKey, Pageable pageable);

    /**
     * NEW: Search by device ID
     */
    Page<AuditLogResponse> searchByDeviceId(String deviceId, Pageable pageable);

    /**
     * NEW: Advanced search with all filters
     */
    Page<AuditLogResponse> advancedSearch(
            String entityType,
            String action,
            String username,
            String licenseKey,
            String deviceId,
            LocalDateTime startDate,
            LocalDateTime endDate,
            Pageable pageable);
}

File: licensetracker_frontend/src/services/auditLogService.js
// src/services/auditLogService.js
import api from './api';

const auditLogService = {
    /**
     * Get all audit logs
     */
    getAllAuditLogs: async (page = 0, size = 20) => {
        const response = await api.get(`/audit-logs?page=${page}&size=${size}`);
        return response.data;
    },

    /**
     * Get audit log by ID
     */
    getAuditLogById: async (id) => {
        const response = await api.get(`/audit-logs/${id}`);
        return response.data;
    },

    /**
     * Get audit logs by user
     */
    getAuditLogsByUser: async (userId, page = 0, size = 20) => {
        const response = await api.get(`/audit-logs/user/${userId}?page=${page}&size=${size}`);
        return response.data;
    },

    /**
     * Get audit logs by entity
     */
    getAuditLogsByEntity: async (entityType, entityId, page = 0, size = 20) => {
        const response = await api.get(
            `/audit-logs/entity/${entityType}/${entityId}?page=${page}&size=${size}`
        );
        return response.data;
    },

    /**
     * Filter audit logs
     */
    filterAuditLogs: async (filters, page = 0, size = 20) => {
        const params = new URLSearchParams();
        params.append('page', page);
        params.append('size', size);
        
        if (filters.entityType) params.append('entityType', filters.entityType);
        if (filters.action) params.append('action', filters.action);
        if (filters.userId) params.append('userId', filters.userId);
        if (filters.startDate) params.append('startDate', filters.startDate);
        if (filters.endDate) params.append('endDate', filters.endDate);

        const response = await api.get(`/audit-logs/filter?${params}`);
        return response.data;
    },

    /**
     * Search audit logs
     */
    searchAuditLogs: async (searchTerm, page = 0, size = 20) => {
        const response = await api.get(
            `/audit-logs/search?searchTerm=${searchTerm}&page=${page}&size=${size}`
        );
        return response.data;
    },

    /**
     * NEW: Search by license key
     */
    searchByLicenseKey: async (licenseKey, page = 0, size = 20) => {
        const response = await api.get(
            `/audit-logs/search/license-key?licenseKey=${encodeURIComponent(licenseKey)}&page=${page}&size=${size}`
        );
        return response.data;
    },

    /**
     * NEW: Search by device ID
     */
    searchByDeviceId: async (deviceId, page = 0, size = 20) => {
        const response = await api.get(
            `/audit-logs/search/device-id?deviceId=${encodeURIComponent(deviceId)}&page=${page}&size=${size}`
        );
        return response.data;
    },

    /**
     * NEW: Advanced search with all filters
     */
    advancedSearch: async (filters, page = 0, size = 20) => {
        const params = new URLSearchParams();
        params.append('page', page);
        params.append('size', size);
        
        if (filters.entityType) params.append('entityType', filters.entityType);
        if (filters.action) params.append('action', filters.action);
        if (filters.username) params.append('username', filters.username);
        if (filters.licenseKey) params.append('licenseKey', filters.licenseKey);
        if (filters.deviceId) params.append('deviceId', filters.deviceId);
        if (filters.startDate) params.append('startDate', filters.startDate);
        if (filters.endDate) params.append('endDate', filters.endDate);

        const response = await api.get(`/audit-logs/search/advanced?${params}`);
        return response.data;
    }
};

export default auditLogService;
